### PR TITLE
rpm_ostree_pkg: new module

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY

Added functionality to manage overlay packages using
rpm-ostree command.

Signed-off-by: Dusty Mabe <dusty@dustymabe.com>
Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
changelogs/fragments/rpm_ostree_package_manager.yml
plugins/modules/packaging/os/rpm_ostree_pkg.py

### CODE REVIEW

**Comment by Akasurde** [comment]
cc @dustymabe @felixfontein @Andersson007 @gundalow Please provide feedback. I'm working on integration tests will provide once those are ready. Thanks in advance.

File: plugins/modules/packaging/os/rpm_ostree_pkg.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+__metaclass__ = type
+
+
+DOCUMENTATION = r'''
+---
+module: rpm_ostree_pkg
+short_description: Install / Uninstall overlay additional packages
+version_added: "2.10"
+description:
+    - This module can be used to install / uninstall overlay additional packages using rpm-ostree command.
```
Comment (Andersson007) [inline plugins/modules/packaging/os/rpm_ostree_pkg.py:17]:
```suggestion
    - Install or uninstall overlay additional packages using rpm-ostree command.
```

File: plugins/modules/packaging/os/rpm_ostree_pkg.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    name:
+      description:
+      - Name of overlay package to install / remove.
+      required: true
+      type: list
+      elements: str
+      aliases: [ pkg ]
+    state:
+      description:
+      - state of the overlay package.
```
Comment (Andersson007) [inline plugins/modules/packaging/os/rpm_ostree_pkg.py:28]:
```suggestion
      - State of the overlay package.
```

File: plugins/modules/packaging/os/rpm_ostree_pkg.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+      - C(present) and C(installed) will simply ensure that a desired package is installed.
+      - C(absent) and C(removed) will remove the specified package.
+      choices: [ 'absent', 'installed', 'present', 'removed' ]
```
Comment (Andersson007) [inline plugins/modules/packaging/os/rpm_ostree_pkg.py:31]:
```suggestion
      choices: [ 'absent', 'present' ]
```
If `absent` and `removed` do the same, imo better not to use extra aliases

File: plugins/modules/packaging/os/rpm_ostree_pkg.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+        # Decide action to perform
+        if self.state in ('installed', 'present'):
+            results['action'] = 'install'
+            cmd.append('install')
+        elif self.state in ('absent', 'removed'):
+            results['action'] = 'uninstall'
+            cmd.append('uninstall')
+
+        # additional parameters
```
Comment (Andersson007) [inline plugins/modules/packaging/os/rpm_ostree_pkg.py:120]:
```suggestion
        # Additional parameters
```

File: plugins/modules/packaging/os/rpm_ostree_pkg.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+        self.module.exit_json(**results)
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            state=dict(
+                default="present",
+                choices=['installed', 'removed', 'absent', 'present']
```
Comment (Andersson007) [inline plugins/modules/packaging/os/rpm_ostree_pkg.py:157]:
```suggestion
                choices=['absent', 'present']
```

File: plugins/modules/packaging/os/rpm_ostree_pkg.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        for pkg in self.params['name']:
+            cmd.append(pkg)
+            results['packages'].append(pkg)
+
+        rc, out, err = (0, '', '')
+        if not self.module.check_mode:
+            rc, out, err = self.module.run_command(cmd)
+
+        results.update(dict(
+            rc=rc,
```
Comment (Andersson007) [inline plugins/modules/packaging/os/rpm_ostree_pkg.py:131]:
```suggestion
        rc, out, err = (0, '', '')
        if not self.module.check_mode:
            rc, out, err = self.module.run_command(cmd)

        results.update(dict(
            rc=rc,
```
but if it's already installed, the module in check_mode will return changed=True anyway?
can we check anyhow whether a package is already installed?

**Comment by Andersson007** [comment]
Two more points:
- AFAIK changelog fragments are not needed for new modules.
- And would be cool to have CI tests as well.

File: changelogs/fragments/rpm_ostree_package_manager.yml [hunk 7]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+- Added a functionality to manage overlay packages using rpm_ostree command.
```
Comment (felixfontein) [inline changelogs/fragments/rpm_ostree_package_manager.yml:2]: Changelog fragments are not needed for new modules and plugins (except `filter` and `test` plugins).

File: plugins/modules/packaging/os/rpm_ostree_pkg.py (hunk 1 above)
Comment (Akasurde) [reply plugins/modules/packaging/os/rpm_ostree_pkg.py:17]: Done.

File: plugins/modules/packaging/os/rpm_ostree_pkg.py (hunk 6 above)
Comment (Akasurde) [reply plugins/modules/packaging/os/rpm_ostree_pkg.py:131]: rpm-ostree provides `--idempotent` parameter which checks it already installed packages for us and returns err code 77 for unchanged.

Comment (Andersson007) [reply plugins/modules/packaging/os/rpm_ostree_pkg.py:131]: ok, in check_mode it won't be checked, will it?

Comment (Andersson007) [reply plugins/modules/packaging/os/rpm_ostree_pkg.py:131]: check_mode will always return changed=True, yes?

Comment (Akasurde) [reply plugins/modules/packaging/os/rpm_ostree_pkg.py:131]: I am afraid, yes, since there is no command to query existing package (atleast I didn't find).

**Comment by felixfontein** [comment]
To fix the sanity test problem, you need to create a symlink from `plugins/modules/rpm_ostree_pkg.py` to the real module file.

File: plugins/modules/packaging/os/rpm_ostree_pkg.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+__metaclass__ = type
+
+
+DOCUMENTATION = r'''
+---
+module: rpm_ostree_pkg
+short_description: Install or Uninstall overlay additional packages
+version_added: "2.10"
+description:
+    - Install or uninstall overlay additional packages using rpm-ostree command.
```
Comment (felixfontein) [inline plugins/modules/packaging/os/rpm_ostree_pkg.py:17]:
```suggestion
    - Install or uninstall overlay additional packages using C(rpm-ostree) command.
```

File: plugins/modules/packaging/os/rpm_ostree_pkg.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+
+- name: Remove overlay package
+  community.general.rpm_ostree_pkg:
+    name: nfs-utils
+    state: absent
+'''
+
+RETURNS = r'''
+rc:
+    description: Return code of rpm-ostree command
```
Comment (felixfontein) [inline plugins/modules/packaging/os/rpm_ostree_pkg.py:53]:
```suggestion
    description: Return code of rpm-ostree command.
```
(Same for the others.)

File: plugins/modules/packaging/os/rpm_ostree_pkg.py (hunk 6 above)
Comment (felixfontein) [reply plugins/modules/packaging/os/rpm_ostree_pkg.py:131]: Maybe in that case the module should not advertise it supports check mode?

**Comment by dustymabe** [comment]
I've been busy but will try to look at this soon

**Comment by Andersson007** [comment]
any updates on this?

File: plugins/modules/packaging/os/rpm_ostree_pkg.py [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+
+DOCUMENTATION = r'''
+---
+module: rpm_ostree_pkg
+short_description: Install or Uninstall overlay additional packages
+version_added: "2.10"
```
Comment (felixfontein) [inline plugins/modules/packaging/os/rpm_ostree_pkg.py:15]:
```suggestion
version_added: 2.0.0
```

File: plugins/modules/packaging/os/rpm_ostree_pkg.py [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+
+DOCUMENTATION = r'''
+---
+module: rpm_ostree_pkg
+short_description: Install or Uninstall overlay additional packages
```
Comment (samccann) [inline plugins/modules/packaging/os/rpm_ostree_pkg.py:14]:
```suggestion
short_description: Install or uninstall overlay additional packages
```

File: plugins/modules/packaging/os/rpm_ostree_pkg.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+    name: nfs-utils
+    state: present
+
+- name: Remove overlay package
+  community.general.rpm_ostree_pkg:
+    name: nfs-utils
+    state: absent
+'''
+
+RETURNS = r'''
```
Comment (Andersson007) [inline plugins/modules/packaging/os/rpm_ostree_pkg.py:51]:
```suggestion
RETURN = r'''
```

File: plugins/modules/packaging/os/rpm_ostree_pkg.py [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+      elements: str
+      aliases: [ pkg ]
+    state:
+      description:
+      - State of the overlay package.
+      - C(present) will simply ensure that a desired package is installed.
+      - C(absent) will remove the specified package.
+      choices: [ 'absent', 'present' ]
+      default: 'present'
+      type: str
```
Comment (Andersson007) [inline plugins/modules/packaging/os/rpm_ostree_pkg.py:33]:
```suggestion
      type: str
notes:
- Supports C(check_mode).
```

File: plugins/modules/packaging/os/rpm_ostree_pkg.py [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+      - Name of overlay package to install / remove.
+      required: true
+      type: list
+      elements: str
+      aliases: [ pkg ]
+    state:
+      description:
+      - State of the overlay package.
+      - C(present) will simply ensure that a desired package is installed.
+      - C(absent) will remove the specified package.
```
Comment (Andersson007) [inline plugins/modules/packaging/os/rpm_ostree_pkg.py:30]:
```suggestion
      - C(present) simply ensures that a desired package is installed.
      - C(absent) removes the specified package.
```
Use the present tense wherever possible (this is recommended by the Ansible documentation).

**Comment by gundalow** [comment]
needs_info

**Review by samccann — APPROVED** [review]
docs portion LGTM

File: plugins/modules/packaging/os/rpm_ostree_pkg.py [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+                choices=['absent', 'present']
+            ),
+            name=dict(
+                aliases=["pkg"],
+                required=True,
+                type='list',
+                elements='str',
+            ),
+        ),
+        supports_check_mode=True
```
Comment (felixfontein) [inline plugins/modules/packaging/os/rpm_ostree_pkg.py:168]:
```suggestion
        supports_check_mode=False
```
from the discussion above

**Comment by felixfontein** [comment]
The PR also needs tests.

File: plugins/modules/packaging/os/rpm_ostree_pkg.py [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+      description:
+      - State of the overlay package.
+      - C(present) simply ensures that a desired package is installed.
+      - C(absent) removes the specified package.
+      choices: [ 'absent', 'present' ]
+      default: 'present'
+      type: str
+author:
+- Dusty Mabe (@dustymabe)
+- Abhijeet Kasurde (@Akasurde)
```
Comment (felixfontein) [inline plugins/modules/packaging/os/rpm_ostree_pkg.py:36]:
```suggestion
- Abhijeet Kasurde (@Akasurde)
notes:
- Does not support C(check_mode).
```

Comment (Akasurde) [reply plugins/modules/packaging/os/rpm_ostree_pkg.py:36]: isn't this by default.

Comment (felixfontein) [reply plugins/modules/packaging/os/rpm_ostree_pkg.py:36]: Most modules simply don't specify, and I think more people assume that check mode is supported by default. So I think it's better to be explicit.

**Review by felixfontein — APPROVED** [review]

**Review by Andersson007 — APPROVED** [review]

**Comment by Andersson007** [comment]
@Akasurde @dustymabe thanks for the module!
@felixfontein @gundalow @samccann ansibullbot thanks for reviewing!

**Comment by Akasurde** [comment]
Thank god I finished this within a year 🤣

**Pull request merged**